### PR TITLE
468 clustering refactor from occurrence

### PR DESCRIPTION
**Pull request by timrobertson100**
This ports the clustering from occurrence into pipelines according to #468 (and relates to #467).

The intention here is to provide common core classes that are suitable for sharing in implementations in GBIF and ALA. The GBIF implementation provided here is a spark-based pipeline, which is ported from the occurrence project. The ALA implementation is a Beam approach and is a WIP on https://github.com/gbif/pipelines/pull/467. I expect in the future these may be aligned further, but this is not the intention now.

@MattBlissett has reviewed the code in occurrence and it is in operation, so the review here is for suitability for the pipelines project, and less about the actual algorithm which will evolve in this project.

CC @djtfmartin for info.

### CODE REVIEW

File: sdks/core/src/test/java/org/gbif/pipelines/core/parsers/clustering/OccurrenceFeaturesPojo.java [hunk 1]
```diff
@@ -1,75 +1,34 @@
...
-    this.recordedBy = recordedBy;
-    this.fieldNumber = fieldNumber;
-    this.recordNumber = recordNumber;
-    this.catalogNumber = catalogNumber;
-    this.otherCatalogNumbers = otherCatalogNumbers;
-  }
+  private final String id;
+  private final String datasetKey;
+  private final Integer speciesKey;
+  private final Integer taxonKey;
```
Comment (djtfmartin) [inline sdks/core/src/test/java/org/gbif/pipelines/core/parsers/clustering/OccurrenceFeaturesPojo.java:15]: Sorry to ask but is a requirement that speciesKey and taxonKey be Integers ?

**Review by muttcg — APPROVED** [review]
For time-saving, I made cosmetic changes here:
https://github.com/gbif/pipelines/pull/483/commits/82daf583fe0b58271640aeb2389451ae523b6e20

File: sdks/core/src/test/java/org/gbif/pipelines/core/parsers/clustering/OccurrenceFeaturesPojo.java (hunk 1 above)
Comment (djtfmartin) [reply sdks/core/src/test/java/org/gbif/pipelines/core/parsers/clustering/OccurrenceFeaturesPojo.java:15]: After discussion with @timrobertson100 ive raised this issue: https://github.com/gbif/pipelines/issues/484

**Comment by timrobertson100** [comment]
Thanks a million for cleaning this up @muttcg.
Can you merge it if you're ok with it, please?

I'll change the integer to string ID stuff in a separate PR tomorrow.

**Comment by timrobertson100** [comment]
Merging following confirmation from @muttcg in private comms

**Pull request merged**